### DISCUSSION TITLE
Customizable synthetization

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#724] [Language design]
I have an idea for a feature but no syntax for it.

Because metatypes will enable some level of reflective programming, it should be possible to use them to express how a trait conformance can be synthesized using the runtime API, rather than hardcoding synthesizers for pre-defined traits in the compiler.

The challenge is that, for an arbitrary type, there's no way to use the type system to guarantee that all parts of a type satisfy a desired property (e.g., to synthesize `Equatable` for lambda types). So we would probably need a way to programmatically test if a type satisfies these properties.

It seems reasonable to imagine that such a test could be executed as a compile-time function.

**Comment by dabrahams** [comment]
It's worth looking at [this pitch](https://forums.swift.org/t/automatic-requirement-satisfaction-in-plain-swift/37158), which accomplishes the same goal.
